### DISCUSSION TITLE
Advanced Dependency Question for testing: Must it always be an instance of a class for dependency injection?

### DISCUSSION THREAD

**Discussion by Forsyth-Creations** [fastapi/fastapi#11690] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#----------------------------------- test_items.py --------------------------

import json

# From the common models, import get_mongo
from app.common_models import (
    create_access_token,
    AppAndMongo
)
from colorama import Fore
from bson.objectid import ObjectId

def generate_fake_token():
    return create_access_token(
        {
            "sub": "some_email",
            "email": "some_email",
            "collections": ["test_db"],
        },
        secret_key="test_secret",
    )

def get_example_json(file_path: str) -> dict:
    with open(file_path, "r") as file:
        return json.loads(file.read())


# Test the function get_items using the TestClient
def test_add_items(basic_client_user_init: AppAndMongo):
    # Print out all the collections in the database
    # mongo = initialize_db
    # print(mongo.list_collection_names())

    # Insert some items in the database
    item = get_example_json("tests/examples/items.json")
    fake = generate_fake_token()

    # Insert an item into the testing database (/items/add)
    print(f"{Fore.GREEN}Inserting item into the database")
    print(fake)
    response = basic_client_user_init.app.post(
        "/items/add",
        json=item,
        headers={"Authorization": f"Bearer {fake}", "collection": "test_db"},
    )

    # Check if the item made it into the database
    mongo = basic_client_user_init.mongo
    uid = response.json()["uid"]

    # Print out all the items
    print(f"{Fore.GREEN}Items in the database")
    print(list(mongo["test_db"]["items"].find()))

    # Check that the item was inserted into the database
    mongo_db = mongo["test_db"]

    print("Checking if the item was inserted into the database")
    print(f"UID: {uid}")

    item = mongo_db["items"].find_one({"_id": ObjectId(uid)})
    assert item != None


# ------------- items.py -----------------------------------
an_instance = EnsurePermissions([add_item])

@router.post(
    "/add",
    dependencies=[
        Depends(an_instance)
    ],  # , Depends(verify_user)
)
def add_a_new_item(
    item: PaduaItem,
    my_mongo=Depends(get_mongo),
    collection=Depends(get_collection_from_headers),
):
    """Add an to the database into the items table"""
    db = my_mongo[collection]
    collection = db["items"]
    final_item = dict(item)

# ------------------------- common.py

class EnsurePermissions:
    def __init__(self, requiredPermissions: List[Power]):
        self.requiredPermissions = requiredPermissions

    def __call__(
        self,
        token: str = Depends(get_token_from_headers),
        database: str = Depends(get_collection_from_headers),
        my_mongo: MongoClient = Depends(get_mongo),
        secret_key: str = Depends(get_secret_key),
    ):
   # more code here



```

### Description

I'm trying to do dependency injection with an advanced dependency (so that my class can define access). Anyway, in the documentation [here](https://fastapi.tiangolo.com/advanced/advanced-dependencies/) I see it needs to be an instance of the class to work. Is there a way to just override the class so I don't have to keep making instances?

### Operating System

Linux

### Operating System Details

Running Linux, but server is running in Docker with the python:3.10-slim image

### FastAPI Version

0.111.0

### Pydantic Version

2.7.3

### Python Version

3.10

### Additional Context

Testing uses a mongomock as another dependency for better testing

**Comment by PhysicallyActive** [comment, marked as answer]
I'm a bit confused what you're asking. If you want to (i) use a class as a dependency and (ii) only use one instance of that class (i.e., never initialising more than one instance), you could use a singleton-pattern for your class to return the same class instance over and over.

**Reply by Forsyth-Creations** [reply]
I can give that a shot. Then the dependency override would override the class/any subsequent instance?

**Reply by PhysicallyActive** [reply]
It doesn't override it, instead it would return the same instance over and over again.
